Applications/YOLO/jni/yolo_v2_loss.cpp

```diff
@@ -0,0 +1,432 @@
+/**
+ * Copyright (C) 2023 Hyeonseok Lee <hs89.lee@samsung.com>
+ *
+ * @file   yolo_v2_loss.cpp
+ * @date   07 March 2023
+ * @brief  This file contains the yolo v2 loss layer
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Hyeonseok Lee <hs89.lee@samsung.com>
+ * @bug    No known bugs except for NYI items
+ *
+ */
+
+#include "yolo_v2_loss.h"
+
+namespace custom {
+
+static constexpr size_t SINGLE_INOUT_IDX = 0;
+
+enum YoloV2LossParams {
+  bbox_x_pred,
+  bbox_y_pred,
+  bbox_w_pred,
+  bbox_h_pred,
+  confidence_pred,
+  class_pred,
+  bbox_w_pred_anchor,
+  bbox_h_pred_anchor,
+  bbox_x_gt,
+  bbox_y_gt,
+  bbox_w_gt,
+  bbox_h_gt,
+  confidence_gt,
+  class_gt,
+  bbox_class_mask,
+  iou_mask
+};
+
+namespace props {
+MaxObjectNumber::MaxObjectNumber(const unsigned &value) { set(value); }
+ClassNumber::ClassNumber(const unsigned &value) { set(value); }
+GridHeightNumber::GridHeightNumber(const unsigned &value) { set(value); }
+GridWidthNumber::GridWidthNumber(const unsigned &value) { set(value); }
+ImageHeightSize::ImageHeightSize(const unsigned &value) { set(value); }
+ImageWidthSize::ImageWidthSize(const unsigned &value) { set(value); }
+} // namespace props
+
+float mse(nntrainer::Tensor &t1, nntrainer::Tensor &t2) {
+  nntrainer::Tensor residual;
+  t1.subtract(t2, residual);
+
+  float l2norm = residual.l2norm();
+  l2norm *= l2norm / residual.size();
+
+  return l2norm;
+}
+
+void msePrime(nntrainer::Tensor &pred, nntrainer::Tensor &ground_truth,
+              nntrainer::Tensor &outgoing_derivative) {
+  pred.subtract(ground_truth, outgoing_derivative);
+  float divider = ((float)pred.size()) / 2;
+  if (outgoing_derivative.divide_i(divider) != ML_ERROR_NONE) {
+    throw std::runtime_error(
+      "[YoloV2LossLayer::calcDerivative] Error when calculating loss");
+  }
+}
+
+YoloV2LossLayer::YoloV2LossLayer() :
+  anchors_w({1, 1, NUM_ANCHOR, 1}, anchors_w_buf),
+  anchors_h({1, 1, NUM_ANCHOR, 1}, anchors_h_buf),
+  sigmoid(nntrainer::ActivationType::ACT_SIGMOID, false),
+  softmax(nntrainer::ActivationType::ACT_SOFTMAX, false),
+  yolo_v2_loss_props(props::MaxObjectNumber(), props::ClassNumber(),
+                     props::GridHeightNumber(), props::GridWidthNumber(),
+                     props::ImageHeightSize(), props::ImageWidthSize()) {
+  anchors_ratio = anchors_w.divide(anchors_h);
+  wt_idx.fill(std::numeric_limits<unsigned>::max());
+}
+
+void YoloV2LossLayer::finalize(nntrainer::InitLayerContext &context) {
+  nntrainer::TensorDim input_dim =
+    context.getInputDimensions()[SINGLE_INOUT_IDX];
+  const unsigned int batch_size = input_dim.batch();
+  const unsigned int class_number =
+    std::get<props::ClassNumber>(yolo_v2_loss_props).get();
+  const unsigned int grid_height_number =
+    std::get<props::GridHeightNumber>(yolo_v2_loss_props).get();
+  const unsigned int grid_width_number =
+    std::get<props::GridWidthNumber>(yolo_v2_loss_props).get();
+  const unsigned int max_object_number =
```

Comment:
I wonder what ```max_object_number``` means. Is it the maximum object number in a picture through out the whole dataset ?